wdl-grammar/src/experimental/lexer.rs

```diff
@@ -0,0 +1,240 @@
+    pub fn peek(&mut self) -> Option<(LexerResult<T>, SourceSpan)> {
+        *self.peeked.get_or_insert_with(|| {
+            self.lexer
+                .next()
+                .map(|r| (r, to_source_span(self.lexer.span())))
+        })
+    }
+
+    /// Morph this lexer into a lexer for a new token type.
+    ///
+    /// The returned lexer continues to point at the same span
+    /// as the current lexer.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the current lexer was peeked without
+    /// consuming the peeked token.
+    pub fn morph<T2>(self) -> Lexer<'a, T2>
+    where
+        T2: Logos<'a, Source = T::Source>,
+        T::Extras: Into<T2::Extras>,
+    {
+        assert!(
+            self.peeked.is_none(),
+            "cannot morph a lexer without consuming a peeked token"
+        );
+
+        Lexer {
+            lexer: self.lexer.morph(),
+            peeked: None,
+        }
+    }
+}
+
+impl<'a, T> Iterator for Lexer<'a, T>
+where
+    T: Logos<'a, Source = str, Error = Error>,
+{
+    type Item = (LexerResult<T>, SourceSpan);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(peeked) = self.peeked.take() {
+            peeked
+        } else {
+            self.lexer
+                .next()
+                .map(|r| (r, to_source_span(self.lexer.span())))
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use pretty_assertions::assert_eq;
+
+    use super::*;
+
+    pub(crate) fn map<T>(
+        (t, s): (LexerResult<T>, SourceSpan),
+    ) -> (LexerResult<T>, std::ops::Range<usize>) {
+        (t, s.offset()..s.offset() + s.len())
+    }
+
+    #[test]
+    fn test_version_1_0() {
+        use VersionToken::*;
+        let lexer = Lexer::<VersionToken>::new(
+            "
+# Test for 1.0 documents
+version 1.0",
+        );
+        let tokens: Vec<_> = lexer.map(map).collect();
+        assert_eq!(
+            tokens,
+            &[
+                (Ok(Whitespace), 0..1),
+                (Ok(Comment), 1..25),
+                (Ok(Whitespace), 25..26),
+                (Ok(VersionKeyword), 26..33),
+                (Ok(Whitespace), 33..34),
+                (Ok(Version), 34..37)
+            ],
+            "produced tokens did not match the expected set"
+        );
+    }
+
+    #[test]
+    fn test_version_1_1() {
+        use VersionToken::*;
+        let lexer = Lexer::<VersionToken>::new(
+            "
+# Test for 1.1 documents
+version 1.1",
+        );
+        let tokens: Vec<_> = lexer.map(map).collect();
+        assert_eq!(
+            tokens,
+            &[
+                (Ok(Whitespace), 0..1),
+                (Ok(Comment), 1..25),
+                (Ok(Whitespace), 25..26),
+                (Ok(VersionKeyword), 26..33),
+                (Ok(Whitespace), 33..34),
+                (Ok(Version), 34..37)
+            ],
+            "produced tokens did not match the expected set"
+        );
+    }
+
+    #[test]
+    fn test_version_draft3() {
```

Comment:
Someone can correct me if I'm wrong here, but I think for all of our sanity we only want to support 1.X versions of WDL. The drafts are all relatively ancient now, and I don't think much of anyone is using them actively.

Reply:
Agreed on all points; this is more of a test of the regex to match a "version" than anything else (definitely not going to support `draft-3` files in the parser implementation). I'm happy to use a different version name to help reduce potential confusion over what we support.

Reply:
FYI: Cromwell just released [v87](https://github.com/broadinstitute/cromwell/releases/tag/87) which retired the `womtool upgrade` command to convert draft-2 to version 1.0. The functionality obviously exists in older womtool versions, but I'd take this as a signal it's safe to stick to 1.x, and agree on the sanity aspect.

Reply:
@peterhuene that makes sense! I think we can probably just add a comment stating the above instead of changing the test.

@rickymagner thanks for the input! Good to get some confirmation on this.